service/history/handler.go

```diff
@@ -2421,6 +2422,54 @@ func (h *Handler) CompleteNexusOperation(ctx context.Context, request *historyse
 	return &historyservice.CompleteNexusOperationResponse{}, nil
 }
 
+func (h *Handler) CompleteNexusOperationChasm(
+	ctx context.Context,
+	request *historyservice.CompleteNexusOperationChasmRequest,
+) (_ *historyservice.CompleteNexusOperationChasmResponse, retErr error) {
+	defer metrics.CapturePanic(h.logger, h.metricsHandler, &retErr)
+	h.startWG.Wait()
+
+	if h.isStopped() {
+		return nil, errShuttingDown
+	}
+
+	ref := chasm.ProtoRefToComponentRef(request.Completion.Ref)
+	info := &persistencespb.ChasmNexusCompletionInfo{
+		CloseTime: request.CloseTime,
+		RequestId: request.Completion.RequestId,
+	}
+	switch variant := request.Outcome.(type) {
+	case *historyservice.CompleteNexusOperationChasmRequest_Failure:
+		info.Outcome = &persistencespb.ChasmNexusCompletionInfo_Failure{
+			Failure: variant.Failure,
+		}
+	case *historyservice.CompleteNexusOperationChasmRequest_Success:
+		info.Outcome = &persistencespb.ChasmNexusCompletionInfo_Success{
+			Success: variant.Success,
+		}
+	default:
+		return nil, serviceerror.NewUnimplemented("unhandled Nexus operation outcome")
+	}
+
+	// Attempt to access the component and call its invocation method. We execute
+	// this similarly as we would a pure task (holding an exclusive lock), as the
+	// assumption is that the accessed component will be recording (or generating a
+	// task) based on this result.
+	_, err := h.chasmEngine.UpdateComponent(ctx, ref, func(ctx chasm.MutableContext, component chasm.Component) error {
```

Comment:
Use `chasm.UpdateComponent` and avoid casting. If you merge the current main changes, you'll see that the chasm engine should already be on the request context.